### PR TITLE
[Builder] Fix pushing images to configured registry with a specific repository

### PR DESCRIPTION
**Pull request by quaark**
Fixes - https://jira.iguazeng.com/browse/ML-2404 

This bug affected the `Spark3Runtime.deploy_default_image` method (but could also affect every other mlrun image) where if a docker registry was configured with a specific repository (e.g. `index.docker.io/quaark` instead of `index.docker.io`) the image would be pushed without the repository in the image name.

Now we push all the images to the configured repository and only if the repository wasn't configured we push to the registry directly.

### CODE REVIEW

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Well done! lets add a test for that use case

File: mlrun/builder.py [hunk 1]
```diff
@@ -538,3 +527,30 @@ def _generate_builder_env(project, builder_env):
     for key, value in builder_env.items():
         env.append(client.V1EnvVar(name=key, value=value))
     return env
+
+
+def _get_image_dest_and_registry_secret(
```
Comment (Tankilevitch) [inline mlrun/builder.py:532]:
```suggestion
def _resolve_image_dest_and_registry_secret(
```

File: mlrun/builder.py [hunk 2]
```diff
@@ -538,3 +527,30 @@ def _generate_builder_env(project, builder_env):
...
+    dest: str, registry: str = None, secret_name: str = None
+) -> (str, str):
+    if registry:
+        return "/".join([registry, dest]), secret_name
+
+    if dest.startswith(IMAGE_NAME_ENRICH_REGISTRY_PREFIX):
```
Comment (Tankilevitch) [inline mlrun/builder.py:538]: Lets add a comment near `IMAGE_NAME_ENRICH_REGISTRY_PREFIX` explaining what its purpose

File: tests/test_builder.py [hunk 3]
```diff
@@ -575,6 +575,126 @@ def test_build_runtime_ecr_with_aws_secret(monkeypatch):
...
+            "default.docker.registry/test-image",
+        ),
+    ],
+)
+def test_get_image_dest(dest, registry, default_repository, expected_dest, monkeypatch):
+    docker_registry_secret = "default-docker-registry-secret"
+    monkeypatch.setattr(config.httpdb.builder, "docker_registry", default_repository)
+    monkeypatch.setattr(
+        config.httpdb.builder, "docker_registry_secret", docker_registry_secret
+    )
```
Comment (Tankilevitch) [inline tests/test_builder.py:624]: Is a monkeypatch really required?
```suggestion
def test_get_image_dest(dest, registry, default_repository, expected_dest):
    docker_registry_secret = "default-docker-registry-secret"
    config.httpdb.builder.docker_registry = default_repository
    config.httpdb.builder.docker_registry_secret  = docker_registry_secret
```

File: tests/test_builder.py [hunk 4]
```diff
@@ -575,6 +575,126 @@ def test_build_runtime_ecr_with_aws_secret(monkeypatch):
...
+    ],
+)
+def test_get_registry_secret(
+    dest, registry, secret_name, default_secret_name, expected_secret_name, monkeypatch
+):
+    docker_registry = "default.docker.registry/default-repository"
+    monkeypatch.setattr(config.httpdb.builder, "docker_registry", docker_registry)
+    monkeypatch.setattr(
+        config.httpdb.builder, "docker_registry_secret", default_secret_name
+    )
```
Comment (Tankilevitch) [inline tests/test_builder.py:697]: same

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

**Review by Tankilevitch — APPROVED** [review]

**Review by Tankilevitch — APPROVED** [review]

**Review by Hedingber — APPROVED** [review]

**Pull request merged**